src/common/hoistFunctions.ts

```diff
@@ -0,0 +1,54 @@
+import { NodePath, Visitor } from 'babel-traverse';
+import * as t from 'babel-types';
+import * as babel from 'babel-core';
+
+const gatherFunctionDeclarations: Visitor = {
+  FunctionDeclaration(path: NodePath<t.Function>) {
+    this.decls.push(path.node);
+    path.replaceWith(t.emptyStatement());
+  },
+  FunctionExpression(path: NodePath<t.Function>) {
+    path.skip();
+  }
+}
+
+function handle(path: NodePath<t.Node>, body: t.Statement[]) {
+  const decls: t.Statement[] = [];
+  path.traverse(gatherFunctionDeclarations, { decls });
+  for (const f of decls) {
+    body.unshift(f);
+  }
+}
+
+const visitor: Visitor = {
+  Program(path: NodePath<t.Program>) {
+    handle(path, path.node.body);
+  },
+  FunctionDeclaration(path: NodePath<t.FunctionDeclaration>) {
+    handle(path, path.node.body.body);
+  },
+  FunctionExpression(path: NodePath<t.FunctionExpression>) {
+    handle(path, path.node.body.body);
+  }
+  
+}
```

Comment:
A better way of doing this (one that doesn't use `path.traverse`) is writing visitor that visits every `FunctionDecl` and `FunctionExpression` and pulls to the top of the closest Function parent or the Program parent. These parents can be found using `path.getParent(p => t.isFunction(p) || t.isProgram(p))`.